libsupport/include/katana/ProgressTracer.h

```diff
@@ -0,0 +1,225 @@
+//     later it will be set to a No-op Tracer, on Fini or the destructor they will call
+//     Close on the global ProgressTracer
+//   This prevents GetProgressTracer() from returning nullptr and ensures Tracers are closed
+//
+//   ProgressScope's will only close their ProgressSpan when their ProgressSpan is the active span
+//   This means that if a user exclusively uses ProgressScopes,
+//   Then parent ProgressSpans will not be closed until their child ProgressSpan's close
+//   i.e.
+//   {
+//      auto scope1 = tracer->StartActiveSpan("1");
+//      auto scope2 = tracer->StartActiveSpan("2");
+//      if (err) { return; }
+//      scope2.Close();
+//      auto scope3 = tracer->StartActiveSpan("3");
+//   }
+//   Always results in scope2's ProgressSpan finishing before scope1's
+//   and scope3's ProgressSpan finishing before scope1's if there is no error
+
+class KATANA_EXPORT ProgressTracer {
+  friend class ProgressSpan;
+
+  void SetSpan(std::shared_ptr<ProgressSpan> span);
+
+protected:
+  ProgressTracer(uint32_t host_id, uint32_t num_hosts)
+      : host_id_(host_id), num_hosts_(num_hosts) {}
+
+public:
+  virtual ~ProgressTracer() = default;
+
+  static std::shared_ptr<ProgressTracer> GetProgressTracer() { return tracer_; }
+  static void SetProgressTracer(std::shared_ptr<ProgressTracer> tracer);
+
+  /// Create a new top level span if one does not already exist.
+  /// Otherwise creates a child span of the current span
+  /// Can specify to not finish the current scope’s span by RAII by
+  /// setting finish_on_close=false
+  ProgressScope StartActiveSpan(
+      const std::string& span_name, bool finish_on_close = true);
+  ProgressScope StartActiveSpan(
+      const std::string& span_name,
+      const std::unique_ptr<ProgressContext>& child_of,
+      bool finish_on_close = true);
+
+  /// Set the current active thread-local span to provided span
+  /// Returns a ProgressScope
+  /// Can specify to not finish the span when the scope goes out of scope
+  /// by RAII by setting finish_on_close=false
+  /// If called with a nullptr span,
+  ///    calls StartActiveSpan("error: nullptr span given to SetActiveSpan", finish_on_close)
+  ProgressScope SetActiveSpan(
+      std::shared_ptr<ProgressSpan> span, bool finish_on_close = true);
+
+  /// Create a new top level span if ignore_active_span=true and
+  /// no child_of value is given
+  /// Otherwise creates a child span of the child_of span or active span
+  /// Should only be used to handle multiple active spans simultaneously
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, bool ignore_active_span = false) = 0;
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name, std::shared_ptr<ProgressSpan> child_of) = 0;
+  virtual std::shared_ptr<ProgressSpan> StartSpan(
+      const std::string& span_name,
+      const std::unique_ptr<ProgressContext>& child_of) = 0;
```

Comment:
Are both this comment and the following referring to line 109, or is this for the unique_ptr?

Reply:
It is a general comment.

If we simply need to refer to an existing object for the duration of a call, `const T&` or `T*`, is more appropriate.

1. `Fn(T* t)` / `Fn(const T& t)`: caller promises to keep `t` live for the duration of the call, callee can access t for the duration of the call
2. `Fn(std::unique_ptr<T> t)`: caller is giving `t` to the callee
3. `Fn(const std::unique_ptr<T>& t)`: redundant with 1 and 3 precludes a `t` that is stack allocated.
4. `Fn(std::shared_ptr<T> t)` / `Fn(const std::shared_ptr<T>& t)`: caller and callee will share responsibility for deallocating `t`. Some people make a subtle distinction between the two forms (will share vs may share) [1], but as a matter of style, it might be simpler to just strengthen everything to `std::shared_ptr<T>`.

[1] https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/

Not knowing the code, I don't know if it is always okay to replace `shared_ptr<T>` with `T*`. We should replace `const std::unique_ptr<T>&` with `const T&` at least.

Reply:
Okay I get it now, thanks